runtime/lua/vim/lsp/util.lua

```diff
@@ -1620,18 +1620,14 @@ function M.open_floating_preview(contents, syntax, opts)
     api.nvim_buf_set_var(bufnr, 'lsp_floating_preview', floating_winnr)
   end
 
-  local augroup_name = ('closing_floating_preview_%d'):format(floating_winnr)
-  local ok =
-    pcall(api.nvim_get_autocmds, { group = augroup_name, pattern = tostring(floating_winnr) })
-  if not ok then
+  local augroup_name = 'nvim_closing_floating_preview'
+  if vim.fn.exists(('#%s'):format(augroup_name)) == 0 then
     api.nvim_create_autocmd('WinClosed', {
       group = api.nvim_create_augroup(augroup_name, {}),
-      pattern = tostring(floating_winnr),
-      callback = function()
-        if api.nvim_buf_is_valid(bufnr) then
+      callback = function(args)
+        if tonumber(args.match) == floating_winnr and api.nvim_buf_is_valid(bufnr) then
```

Comment:
`floating_winnr` is bound to the closure when the autocmd was created. But the autocmd is long-lived. Can we check `vim.b[bufnr].lsp_floating_preview` instead?
```suggestion
        if tonumber(args.match) == vim.b[bufnr].lsp_floating_preview and api.nvim_buf_is_valid(bufnr) then
```

Reply:
yes i also noticed it will make the variable move to heap..

Reply:
> and I think bufnr should not be also captured by closure either ? Can it be stored in w[floating_winnr].buf_hold_win?

oh, right. I thought that was an event arg. We should get it from event args, or `api.nvim_get_current_buf()` or whatever.

Reply:
> We should get it from event args, or api.nvim_get_current_buf() 

1. `args.buf` is current buffer in floating_winnr  
2.  if there trigger BufEnter which will close floating_winnr and get_current_buf will return a wrong buffer which maybe not hold the floating_winnr.

Reply:
either `vim.b[bufnr].lsp_floating_preview` is valid or not. why does any of that matter?

Reply:
> why does any of that matter?

we don't capture bufnr in closure ,then need to find bufnr which holds the floatinng_winnr. The event args or api.nvim_get_current_buf you proposed is not the previous bufnr in some cases, so this is invalid. Ineed to store this bufnr somewhere like win-var

Reply:
this is getting too complicated